internal/cron/cron.go

```diff
@@ -0,0 +1,508 @@
+		result = standardResult
+	}
+	// If Sunday as a value of 7 is allowed by step, check the logic with a value of 0
+	if stepAllowsSecondSunday {
+		if secondSundayResult, ok := nextValueForDayOfStandardizedWeekStep(newStepWithValue(0), values, t); ok {
+			if result == -1 || secondSundayResult < result {
+				result = secondSundayResult
+			}
+		}
+	}
+	return result, result != -1
+}
+
+func nextValueForDayOfStandardizedWeekStep(step Step, values componentValues, t componentType) (int, bool) {
+	// Ignores Sunday == 7
+	start, end, incr := rangeParametersForStep(step, t)
+	if start == 7 {
+		return -1, false
+	}
+	if end == 7 {
+		end = 6
+	}
+
+	valueForCurrentWeekday := max(0, values[dayOfMonth]) // is value is -1, we want day before the current month (ie 0)
+	currentDate := time.Date(values[year], time.Month(values[month]), valueForCurrentWeekday, 0, 0, 0, 0, time.UTC)
+	currentWeekday := int(currentDate.Weekday())
+
+	startOfWeekInMonth := valueForCurrentWeekday - currentWeekday //Sunday
+
+	nextDayOfWeek := -1
+
+	// try current week
+	if currentWeekday < start {
+		nextDayOfWeek = start
+	} else {
+		// round down to the nearest increment from start, then add one increment
+		nextDayOfWeek = start + (((currentWeekday-start)/incr)+1)*incr
+	}
+	if nextDayOfWeek < start || nextDayOfWeek > end {
+		// try next week
+		nextDayOfWeek = 7 + start
+	}
+
+	next := startOfWeekInMonth + nextDayOfWeek
+
+	date := time.Date(values[year], time.Month(values[month]), next, 0, 0, 0, 0, time.UTC)
+	if date.Day() != next {
+		// This month does not not have this day in this particular year (eg Feb 30th)
+		return -1, false
+	}
+	return next, true
+}
+
+func firstValueForComponents(components []Component, values componentValues, t componentType) (int, error) {
+	fakeValues := make([]int, len(values))
+	copy(fakeValues, values)
+	fakeValues[t] = -1
+	return nextValue(components, fakeValues, t)
+}
+
+func timeFromValues(values componentValues) time.Time {
+	return time.Date(values[year],
+		time.Month(values[month]),
+		values[dayOfMonth],
+		values[hour],
+		values[minute],
+		values[second],
+		0,
+		time.UTC)
+}
+
+// Validite makes sure that a pattern has no mistakes in the cron format, and that there is a valid next value from a set point in time
+// Validity checks are done while calculating a next date to ensure that we never calculate a next date for an invalid pattern
+func Validate(pattern Pattern) error {
+	_, err := NextAfter(pattern, time.Date(2020, time.January, 1, 0, 0, 0, 0, time.UTC), true)
+	return err
+}
+
+func validateComponent(component Component, t componentType) error {
+	if len(component.List) == 0 {
+		return fmt.Errorf("%s must have at least value", stringForComponentType(t))
+	}
+	for _, step := range component.List {
+		if step.ValueRange.IsFullRange && (step.ValueRange.Start != nil || step.ValueRange.End != nil) {
+			return fmt.Errorf("range can not have start/end if it is a full range")
+		}
+		if !step.ValueRange.IsFullRange {
```

Comment:
Invert this logic and `continue` to reduce nesting. It makes it reduces nesting and makes the code easier to read.

Reply:
Mm there's a bug there as well, the step increment needs to be validated even it is a full range. Made both changes 👍